src/Datasource/EntityTrait.php

```diff
@@ -242,13 +243,39 @@ public function set($property, $value = null, array $options = [])
         $options += ['setter' => true, 'guard' => $guard];
 
         foreach ($property as $p => $value) {
+            $nested = false;
+            if (strpos($p, '.') > 0 && !array_key_exists($p, $this->_properties)) {
+                list($p, $nested) = explode('.', $p, 2);
+            }
+
             if ($options['guard'] === true && !$this->accessible($p)) {
                 continue;
             }
 
             $this->dirty($p, true);
 
-            if (!array_key_exists($p, $this->_original) &&
+            if ($nested) {
+                if (!array_key_exists($p, $this->_properties)
+                    || is_null($this->_properties[$p])
+                ) {
+                    $value = new \Cake\ORM\Entity([$nested => $value]);
```

Comment:
That's a good point, but the issue is the impedance mismatch between ::getOriginal and ::set(). When you set a dotted path that must be created, if you create the sub-objects as arrays or ArrayObjects, you can't retain `::$_original` data for those new sub items. It' be _possible_ to discard this information, but it will actually make the implementation in set() and getOriginal() more complex. I'm fine with whatever, but I did actually start with ArrayObjects and realized that path was going to be more complicated.